File: 0042-trapping-rain-water/0042-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        vector<int>rightMax(height.size(),0);
        int n=height.size(),mx=0;
        for(int i=n-1;i>=0;i--)
        {
            mx=max(mx,height[i]);
            rightMax[i]=mx;
        }
        mx=height[0];
        int water=0;
        for(int i=1;i<n-1;i++)
        {
            water+=max(0,min(mx,rightMax[i+1])-height[i]);
            mx=max(mx,height[i]);
        }
        return water;
    }
};

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length;
        int[] leftGreater = new int[n];
        for(int i=0;i<n;i++)
            leftGreater[i]=Math.max(height[i],i==0?0:leftGreater[i-1]);
        int[] rightGreater = new int[n];
        for(int i=n-1;i>=0;i--)
            rightGreater[i]=Math.max(height[i],i==n-1?0:rightGreater[i+1]);
        
        int res=0;
        for(int i=1;i<n-1;i++)
        {
            res+=Math.max(Math.min(leftGreater[i],rightGreater[i])-height[i],0);
        }
        return res;
    }
}